File: wrappers/api-clients/python/src/checkrd_api/_pagination.py
"""Pagination iterators.

Mirrors OpenAI / Anthropic shape: list endpoints return a
:class:`SyncPage` (or :class:`AsyncPage` for ``AsyncCheckrd``) that
is *both* a holder for the current page's ``data`` array *and*
iterable across all subsequent pages transparently.

A user can either:

- Iterate (most ergonomic): ``for agent in client.agents.list(): ...``
  — the loop fetches page 2, 3, … as needed.
- Page manually: ``page = client.agents.list(); while page.has_next_page(): page = page.get_next_page()``.

Reference: ``openai-python/src/openai/pagination.py``.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import (
    TYPE_CHECKING,
    AsyncIterator,
    Callable,
    Generic,
    Iterator,
    List,
    Optional,
    TypeVar,
)

if TYPE_CHECKING:
    from ._client import AsyncCheckrd, Checkrd

_T = TypeVar("_T")


@dataclass
class _PageState(Generic[_T]):
    """State kept by both sync and async pages — the page's payload
    plus enough info to fetch the next one if iteration continues."""

    data: List[_T]
    has_more: bool
    next_cursor: Optional[str]
    path: str
    params: dict
    item_decoder: Callable[[dict], _T]


class SyncPage(Generic[_T]):
    """Synchronous cursor-paginated response.

    The class instance returned from a sync ``list()`` call doubles
    as the current page (``page.data``, ``page.has_more``,
    ``page.next_cursor``) and as an iterator that walks every page
    until the cursor exhausts. Most callers just iterate.
    """

    _client: "Checkrd"
    _state: _PageState[_T]

    def __init__(self, client: "Checkrd", state: _PageState[_T]) -> None:
        self._client = client
        self._state = state

    @property
    def data(self) -> List[_T]:
        return self._state.data

    @property
    def has_more(self) -> bool:
        return self._state.has_more

    @property
    def next_cursor(self) -> Optional[str]:
        return self._state.next_cursor

    def has_next_page(self) -> bool:
        return self._state.has_more and self._state.next_cursor is not None

    def get_next_page(self) -> "SyncPage[_T]":
        """Fetch the next page synchronously. Raises :class:`StopIteration`
        when there is no next page."""
        if not self.has_next_page():
            raise StopIteration
        params = dict(self._state.params)
        params["cursor"] = self._state.next_cursor
        return self._client._get_api_list(
            self._state.path,
            params=params,
            item_decoder=self._state.item_decoder,
        )

    def __iter__(self) -> Iterator[_T]:
        page: SyncPage[_T] = self
        while True:
            for item in page.data:
                yield item
            if not page.has_next_page():
                return
            page = page.get_next_page()


class AsyncPage(Generic[_T]):
    """Asynchronous cursor-paginated response. Mirrors :class:`SyncPage`."""

    _client: "AsyncCheckrd"
    _state: _PageState[_T]

    def __init__(self, client: "AsyncCheckrd", state: _PageState[_T]) -> None:
        self._client = client
        self._state = state

    @property
    def data(self) -> List[_T]:
        return self._state.data

    @property
    def has_more(self) -> bool:
        return self._state.has_more

    @property
    def next_cursor(self) -> Optional[str]:
        return self._state.next_cursor

    def has_next_page(self) -> bool:
        return self._state.has_more and self._state.next_cursor is not None

    async def get_next_page(self) -> "AsyncPage[_T]":
        """Fetch the next page. Raises :class:`StopAsyncIteration`
        when there is no next page."""
        if not self.has_next_page():
            raise StopAsyncIteration
        params = dict(self._state.params)
        params["cursor"] = self._state.next_cursor
        return await self._client._get_api_list(
            self._state.path,
            params=params,
            item_decoder=self._state.item_decoder,
        )

    def __aiter__(self) -> AsyncIterator[_T]:
        return self._iter()

    async def _iter(self) -> AsyncIterator[_T]:
        page: AsyncPage[_T] = self
        while True:
            for item in page.data:
                yield item
            if not page.has_next_page():
                return
            page = await page.get_next_page()

File: wrappers/api-clients/javascript/src/pagination.ts
/**
 * Pagination iterator.
 *
 * Mirrors the OpenAI / Anthropic / Stainless TS shape: list
 * methods return a ``PagePromise`` that is awaitable to get the
 * first page, async-iterable to walk every page, and exposes
 * ``hasNextPage()`` / ``getNextPage()`` for manual cursor walking.
 *
 * Usage::
 *
 *     for await (const agent of client.agents.list()) {
 *       console.log(agent.name);  // fetches page 2, 3, … on demand
 *     }
 *
 *     // Or: get just the first page synchronously after await.
 *     const page = await client.agents.list();
 *     for (const agent of page.data) console.log(agent.name);
 *     if (page.hasNextPage()) {
 *       const next = await page.getNextPage();
 *     }
 */

import type { Checkrd } from "./client.js";

/** Wire shape of every paginated response body. */
export interface PaginatedBody<T> {
  data: T[];
  has_more: boolean;
  next_cursor: string | null;
}

export interface PageOptions {
  path: string;
  params: Record<string, unknown>;
  decode: (raw: unknown) => unknown;
}

/**
 * One page of results plus a cursor to the next. Both an array
 * holder (``page.data``) and async-iterable across all pages.
 */
export class Page<T> implements AsyncIterable<T> {
  readonly data: T[];
  readonly hasMore: boolean;
  readonly nextCursor: string | null;
  private readonly client: Checkrd;
  private readonly opts: PageOptions;

  constructor(
    client: Checkrd,
    body: PaginatedBody<T>,
    opts: PageOptions,
  ) {
    this.client = client;
    this.opts = opts;
    this.data = body.data;
    this.hasMore = body.has_more;
    this.nextCursor = body.next_cursor;
  }

  hasNextPage(): boolean {
    return this.hasMore && this.nextCursor !== null;
  }

  async getNextPage(): Promise<Page<T>> {
    if (!this.hasNextPage()) {
      throw new Error("getNextPage() called when no next page exists");
    }
    return this.client._getApiList<T>(this.opts.path, {
      ...this.opts.params,
      cursor: this.nextCursor,
    });
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    let page: Page<T> = this;
    while (true) {
      for (const item of page.data) yield item;
      if (!page.hasNextPage()) return;
      page = await page.getNextPage();
    }
  }
}

/**
 * The promise type returned by every ``list()`` method. Awaitable
 * to get the first :class:`Page`; async-iterable to walk every
 * page transparently.
 *
 * Mirrors OpenAI's ``PagePromise`` — same dual await/iterate
 * affordance.
 */
export class PagePromise<T> implements PromiseLike<Page<T>>, AsyncIterable<T> {
  private readonly inner: Promise<Page<T>>;

  constructor(inner: Promise<Page<T>>) {
    this.inner = inner;
  }

  then<TResult1 = Page<T>, TResult2 = never>(
    onfulfilled?: ((value: Page<T>) => TResult1 | PromiseLike<TResult1>) | undefined | null,
    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | undefined | null,
  ): PromiseLike<TResult1 | TResult2> {
    return this.inner.then(onfulfilled, onrejected);
  }

  catch<TResult = never>(
    onrejected?: ((reason: unknown) => TResult | PromiseLike<TResult>) | undefined | null,
  ): Promise<Page<T> | TResult> {
    return this.inner.catch(onrejected);
  }

  finally(onfinally?: (() => void) | undefined | null): Promise<Page<T>> {
    return this.inner.finally(onfinally);
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    const page = await this.inner;
    yield* page;
  }
}
